test/testutil/tap_bio.c

```diff
@@ -0,0 +1,221 @@
+            BIO_meth_set_ctrl(tap, tap_ctrl);
+            BIO_meth_set_create(tap, tap_new);
+            BIO_meth_set_destroy(tap, tap_free);
+            BIO_meth_set_callback_ctrl(tap, tap_callback_ctrl);
+        }
+    }
+    return tap;
+}
+
+static int tap_new(BIO *b)
+{
+    BIO_TAP_CTX *ctx;
+
+    if ((ctx = OPENSSL_malloc(sizeof(*ctx))) == NULL)
+        return 0;
+    ctx->newline = 1;
+    ctx->leadingspaces = 0;
+    BIO_set_data(b, ctx);
+    BIO_set_init(b, 1);
+    return 1;
+}
+
+static int tap_free(BIO *b)
+{
+    if (b == NULL)
+        return 0;
+    OPENSSL_free(BIO_get_data(b));
+    BIO_set_data(b, NULL);
+    BIO_set_init(b, 0);
+    return 1;
+}
+
+static int tap_read_ex(BIO *b, char *buf, size_t size, size_t *out_size)
+{
+    BIO *next = BIO_next(b);
+    int ret = 0;
+
+    ret = BIO_read_ex(next, buf, size, out_size);
+    BIO_clear_retry_flags(b);
+    BIO_copy_next_retry(b);
+    return ret;
+}
+
+/*
+ * Output a string to the specified bio and return 1 if successful.
+ */
+static int write_string(BIO *b, const char *buf, size_t n)
+{
+    size_t m;
+
+    return BIO_write_ex(b, buf, n, &m) != 0 && m == n;
+}
+
+/*
+ * Output the specified number of spaces to the bio and return 1 if successful.
+ */
+static int spaces(BIO *b, size_t n)
+{
+    size_t i;
+
+    for (i = 0; i < n; i++)
+        if (!write_string(b, " ", 1))
+            return 0;
+    return 1;
+}
+
+/*
+ * Write some data.
+ *
+ * This function implements a simple state machine that detects new lines.
+ * It indents the output and prefixes it with a '#' character if one isn't
+ * already present.  Indentation after a '#' and at the beginning of a line
+ * without a leading '#' is preserved.
+ *
+ * It returns the number of input characters that were output in in_size.
+ * More characters than this will likely have been output however any calling
+ * code will be unable to correctly assess the actual number of characters
+ * emitted and would be prone to failure if the actual number were returned.
+ */
+static int tap_write_ex(BIO *b, const char *buf, size_t size, size_t *in_size)
+{
+    BIO *next = BIO_next(b);
+    BIO_TAP_CTX *ctx = (BIO_TAP_CTX *)BIO_get_data(b);
+    const size_t indent = subtest_level();
+    const char *p = buf;
+    size_t i, m, n;
+
+    for (i = 0; i < size; i++) {
+        if (ctx->newline) {
+            switch (buf[i]) {
+            case ' ':
+                ctx->leadingspaces++;
+                break;
+
+            case '\t':
+                ctx->leadingspaces = (8 + ctx->leadingspaces) & ~7;
+                break;
+
+            case '\n':
+                ctx->leadingspaces = 0;
+                if (!write_string(next, "\n", 1))
+                    goto err;
+                break;
+
+            case '#':
+                if (!spaces(next, indent))
+                    goto err;
+                p = buf + i;
+                ctx->leadingspaces = ctx->newline = 0;
+                break;
```

Comment:
I think the whole `BIO_write_ex` function could be simplified in light of this.  Essentially, when a \n is seen:

```
write "\n"
indent
write "# "
```

For all other characters, output them immediately.

It wouldn't be that simple, the indent would need to occur when the next character is ready (in case the depth changes betwixt now and then).  However, counting the leading spaces would be unnecessary, the state machine would be simplified and the pointer fun avoided.  I'll give this a go.